kube-runtime/src/cache/mod.rs

```diff
@@ -25,6 +29,87 @@ where
     stream.inspect_ok(move |event| store.apply_watcher_event(event))
 }
 
+/// A simple reflector cache around a store and an owned watcher
+pub struct Cache<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash,
+{
+    reader: Store<K>,
+    cache: BoxStream<'static, std::result::Result<watcher::Event<K>, watcher::Error>>,
+}
+
+impl<K> Cache<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash + Clone + Default,
+{
+    /// Create a Cache on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Cache` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    #[must_use]
+    pub fn new(api: Api<K>, lp: ListParams) -> Self {
+        Self::new_with(api, lp, Default::default())
+    }
+}
+
+impl<K> Cache<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash + Clone,
+{
+    /// Create a Cache on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Cache` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    ///
+    /// This variant constructor is for [`dynamic`] types found through discovery. Prefer [`Cache::new`] for static types.
+    #[must_use]
+    pub fn new_with(api: Api<K>, lp: ListParams, dyntype: K::DynamicType) -> Self {
+        let writer = Writer::<K>::new(dyntype);
+        let reader = writer.as_reader();
+        let cache = reflector(writer, watcher(api, lp)).boxed();
+        Self { reader, cache }
+    }
+
+    /// Retrieve a reading copy of the store before starting the cache
+    #[must_use]
+    pub fn store(&self) -> Store<K> {
+        self.reader.clone()
+    }
+
+    /// Consume the stream and return a future that will run the duration of the program
+    ///
+    /// This should be awaited forever.
+    #[must_use]
+    pub fn run(self) -> BoxFuture<'static, ()> {
+        let stream = self.applies();
+        stream.for_each(|_| futures::future::ready(())).boxed()
+    }
+
+    /// Consumes the cache, runs the reflector, and returns an information stream of watch events (modified/added)
+    ///
+    /// Note that the returned stream is always reflected in the [`reader`](Cache::reader).
+    /// If you do not require a reader, prefer using a [`watcher`] directly.
+    pub fn applies(self) -> impl Stream<Item = K> + Send {
+        utils::try_flatten_applied(self.cache).filter_map(|x| async move { x.ok() })
```

Comment:
We probably don't want to just discard all errors.

Reply:
Well, we advertise the `reflector` / `watcher` as a thing you can just run forever and it should deal with the issues, so I think this plays into how we want to handle #577 . The current `watcher` and `reflector` strategy of just returning a `TryStream` and pretending that the user actually does any greater error handling than inserting a `?` at least crashes the app on errors so that kubernetes can do the backoff itself.

If we instead take a `backoff` object as a parameter to this struct, then this would be a place to catch errors and restart the watcher. However, by simply restarting (with some backoff within this struct  here), we create the potential for a reflector to never initialize in the first place, or start failing continuously from some point (maybe after some rbac changes), and the app would not really know.

I guess my question is, if we are trying to handle #577 ourselves, when/how do we decide to call it quits?

Reply:
Ah, wait. I think I misunderstood you. You want the `backoff` to be baked into the `watcher` itself for this particular error.

Reply:
Carrying on discussions in #577, but now agree that we should not discard errors here.

Reply:
Errors are propagated again in https://github.com/kube-rs/kube-rs/commit/d3473acebb804effd3fc07b5c852430ef16b095f .